contracts/libraries/LockDataLibrary.sol

```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.19;
+
+import {IPoolManager} from "../interfaces/IPoolManager.sol";
+
+library LockDataLibrary {
+    uint256 private constant OFFSET = uint256(keccak256("LockData"));
+
+    function push(IPoolManager.LockData storage self, address locker) internal {
+        uint96 length = self.length;
+        uint256 index = self.index; // only 96 bits, but we expand to 256 for clarity when using bitwise or below
+
+        unchecked {
+            uint256 indexToWrite = OFFSET + length;
+            /// @solidity memory-safe-assembly
+            assembly {
+                // note that the current index becomes the parent lock index
+                sstore(indexToWrite, or(shl(96, locker), index))
+            }
+            self.length = length + 1;
+            self.index = length;
```

Comment:
ok just thinking through this.. if I lock twice with the same address, I get two different entries here, but only one CurrencyDelta shared between both locks. I guess that's OK since we need to unlock twice but only need to ensure zero deltas once?

Reply:
yes that's correct - we need unique locks to maintain the ability to have nested locks etc - the fact that this allows the same locker to get 2 locks in a row is weird, but ultimately as long as we're a) tracking deltas globally by locker + currency and b) only update (global) non-0 delta count when going from 0->non-0 and vice versa, it should be fine, provided that we check that non-0 delta count is 0 after all locks have expired. and the benefit of this is also that you can punt on deltas until your last lock if you know you'll be doing 2 (the primary motivation behind the pr in the first place). so i think this checks out?